src/input/type_base.cc

```diff
@@ -149,17 +168,21 @@ TypeBase::TypeHash Array::content_hash() const
 }
 
 
-bool Array::finish() {
-	return data_->finish();
+bool Array::finish(bool is_generic) {
+	return data_->finish(is_generic);
 }
 
 
 
-bool Array::ArrayData::finish()
+bool Array::ArrayData::finish(bool is_generic)
 {
 	if (finished) return true;
 
-	return (finished = type_of_values_->finish() );
+    ASSERT(is_generic || typeid( *type_of_values_ ) != typeid(Parameter),
+    		"Finished non-generic Array can't be of type Parameter '%s'.\n",
+    		type_of_values_->type_name().c_str());
+
```

Comment:
Remove this assert (as well as similar in Record and AbstractRecord). Replace it by overloaded 
Parameter::finish(...) with an ASSERT(is_generic, ....) reporting actual name of the parameter which is the only helpful information anyway.